File: src/main/pc-agent/types/MessageEnvelope.ts
/**
 * MessageEnvelope — Unified protocol envelope for cross-IDE event streaming.
 *
 * Wraps all messages exchanged between PC Agent, AI-Gateway, and Skill Server
 * to provide consistent metadata, versioning, and sequence tracking.
 */

/**
 * Source identifier for the message origin.
 */
export type MessageSource = 'OPENCODE' | 'CURSOR' | 'WINDSURF';

/**
 * Envelope metadata attached to every protocol message.
 */
export interface EnvelopeMetadata {
  /** Protocol version (semver format, e.g., "1.0.0") */
  version: string;

  /** Unique message identifier (UUID v4) */
  messageId: string;

  /** ISO 8601 timestamp of message creation */
  timestamp: string;

  /** Source IDE/tool that generated this message */
  source: MessageSource;

  /** Agent connection ID (assigned by gateway on registration) */
  agentId: string;

  /** Session identifier (optional, for session-scoped messages) */
  sessionId?: string;

  /** Sequence number for ordering within a scope */
  sequenceNumber: number;

  /** Scope for sequence numbering: 'session' or 'agent' */
  sequenceScope: 'session' | 'agent';
}

/**
 * Complete message envelope with metadata and payload.
 */
export interface MessageEnvelope<T = unknown> {
  /** Envelope metadata */
  envelope: EnvelopeMetadata;

  /** Message type discriminator (e.g., 'tool_event', 'invoke', 'tool_done') */
  type: string;

  /** Message payload (type-specific content) */
  payload: T;
}

/**
 * Type guard to check if a message has an envelope.
 */
export function hasEnvelope(msg: unknown): msg is MessageEnvelope {
  return (
    typeof msg === 'object' &&
    msg !== null &&
    'envelope' in msg &&
    typeof (msg as Record<string, unknown>).envelope === 'object' &&
    'type' in msg &&
    'payload' in msg
  );
}

/**
 * Extract envelope metadata from a message, or return undefined.
 */
export function getEnvelope(msg: unknown): EnvelopeMetadata | undefined {
  return hasEnvelope(msg) ? msg.envelope : undefined;
}

File: skill-server/src/main/java/com/opencode/cui/skill/model/MessageEnvelope.java
package com.opencode.cui.skill.model;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import lombok.Builder;

/**
 * Unified protocol envelope for cross-IDE event streaming.
 *
 * Wraps all messages exchanged between PC Agent, AI-Gateway, and Skill Server
 * to provide consistent metadata, versioning, and sequence tracking.
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@JsonInclude(JsonInclude.Include.NON_NULL)
public class MessageEnvelope {

    /**
     * Envelope metadata attached to every protocol message.
     */
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    @JsonInclude(JsonInclude.Include.NON_NULL)
    public static class EnvelopeMetadata {
        /** Protocol version (semver format, e.g., "1.0.0") */
        private String version;

        /** Unique message identifier (UUID v4) */
        private String messageId;

        /** ISO 8601 timestamp of message creation */
        private String timestamp;

        /** Source IDE/tool that generated this message */
        private String source;

        /** Agent connection ID (assigned by gateway on registration) */
        private String agentId;

        /** Session identifier (optional, for session-scoped messages) */
        private String sessionId;

        /** Sequence number for ordering within a scope */
        private Long sequenceNumber;

        /** Scope for sequence numbering: 'session' or 'agent' */
        private String sequenceScope;
    }

    /** Envelope metadata */
    private EnvelopeMetadata envelope;

    /** Message type discriminator (e.g., 'tool_event', 'invoke', 'tool_done') */
    private String type;

    /** Message payload (type-specific content) */
    private Object payload;

    /**
     * Check if this message has a valid envelope.
     */
    public boolean hasEnvelope() {
        return envelope != null && envelope.getVersion() != null;
    }

    /**
     * Get envelope or return a default with minimal metadata.
     */
    public EnvelopeMetadata getEnvelopeOrDefault() {
        if (hasEnvelope()) {
            return envelope;
        }
        return EnvelopeMetadata.builder()
                .version("0.0.0")
                .messageId("unknown")
                .timestamp("")
                .source("UNKNOWN")
                .agentId("unknown")
                .sequenceNumber(0L)
                .sequenceScope("agent")
                .build();
    }
}
